### PR TITLE
TRUNK-5134-Replace path constructions using separator char

### PR DESCRIPTION
**Pull request by chamathabeysinghe**
## Description of what I changed

Modified 7 files for this task:

- api/src/main/java/org/openmrs/util/UpgradeUtil.java
- api/src/test/java/org/openmrs/obs/BinaryDataHandlerTest.java
- api/src/test/java/org/openmrs/obs/TextHandlerTest.java
- web/src/main/java/org/openmrs/module/web/WebModuleUtil.java
- web/src/main/java/org/openmrs/web/filter/initialization/TestInstallUtil.java
- web/src/test/java/org/openmrs/module/web/WebModuleUtilTest.java
- web/src/test/java/org/openmrs/web/test/WebModuleActivatorTest.java

## Issue I worked on

see https://issues.openmrs.org/browse/TRUNK-5134

## Checklist: I completed these to help reviewers :)

- [x] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [ ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: web/src/main/java/org/openmrs/module/web/WebModuleUtil.java [hunk 1]
```diff
@@ -287,8 +287,8 @@ public static boolean startModule(Module mod, ServletContext servletContext, boo
 			
 			// mark to delete the entire module web directory on exit
 			// this will usually only be used when an improper shutdown has occurred.
-			String folderPath = realPath + "/WEB-INF/view/module/" + mod.getModuleIdAsPath();
-			File outFile = new File(folderPath.replace("/", File.separator));
+			String folderPath = Paths.get(realPath , "WEB-INF", "view", "module", mod.getModuleIdAsPath()).toString();
+			File outFile = new File(folderPath);
```
Comment (teleivo) [inline web/src/main/java/org/openmrs/module/web/WebModuleUtil.java]: can you inline these ? into

```java
File outFile = Paths.get(realPath , "WEB-INF", "view", "module", mod.getModuleIdAsPath()).toFile();
```

File: web/src/main/java/org/openmrs/module/web/WebModuleUtil.java [hunk 2]
```diff
@@ -681,8 +681,8 @@ public static void shutdownModules(ServletContext servletContext) {
 		String realPath = getRealPath(servletContext);
 		
 		// clear the module messages
-		String messagesPath = realPath + "/WEB-INF/";
-		File folder = new File(messagesPath.replace("/", File.separator));
+		String messagesPath = Paths.get(realPath,"WEB-INF").toString();
+		File folder = new File(messagesPath);
```
Comment (teleivo) [inline web/src/main/java/org/openmrs/module/web/WebModuleUtil.java]: same here can you remove the temp variable `messagesPath` by calling `toFile()` on the `Paths.get()`

File: web/src/main/java/org/openmrs/module/web/WebModuleUtil.java [hunk 3]
```diff
@@ -733,8 +733,8 @@ public static void stopModule(Module mod, ServletContext servletContext, boolean
 		String realPath = getRealPath(servletContext);
 		
 		// delete the web files from the webapp
-		String absPath = realPath + "/WEB-INF/view/module/" + moduleId;
-		File moduleWebFolder = new File(absPath.replace("/", File.separator));
+		String absPath = Paths.get(realPath, "WEB-INF", "view", "module", moduleId).toString();
+		File moduleWebFolder = new File(absPath);
```
Comment (teleivo) [inline web/src/main/java/org/openmrs/module/web/WebModuleUtil.java]: same here

File: web/src/main/java/org/openmrs/module/web/WebModuleUtil.java (hunk 1 above)
Comment (chamathabeysinghe) [reply web/src/main/java/org/openmrs/module/web/WebModuleUtil.java]: Sure I can

**Comment by chamathabeysinghe** [comment]
@teleivo I made the requested changes. However test coverage seem to be decreased. Is it needed to write some test cases for that?

**Comment by chamathabeysinghe** [comment]
@teleivo I have removed those File.separators also, but there are some instances File.separator can't be removed.

File: web/src/main/java/org/openmrs/module/web/WebModuleUtil.java [hunk 4]
```diff
@@ -239,8 +240,7 @@ public static boolean startModule(Module mod, ServletContext servletContext, boo
 				if (root.getElementsByTagName("dwr").getLength() > 0) {
 					
 					// get the dwr-module.xml file that we're appending our code to
-					File f = new File(realPath + "/WEB-INF/dwr-modules.xml".replace("/", File.separator));
-					
+					File f = Paths.get(realPath,"WEB-INF","dwr-modules.xml").toFile();
```
Comment (teleivo) [inline web/src/main/java/org/openmrs/module/web/WebModuleUtil.java]: this change is ok, as it does not change the behavior of the code. Since here the replace is used on a string which we control and you correctly cleaned from "/".

File: web/src/main/java/org/openmrs/module/web/WebModuleUtil.java [hunk 5]
```diff
@@ -923,7 +919,7 @@ public static String getModuleWebFolder(String moduleId) {
 		
 		moduleWebFolder += moduleId;
 		
-		return moduleWebFolder.replace("/", File.separator);
+		return Paths.get(moduleWebFolder).toString();
```
Comment (teleivo) [inline web/src/main/java/org/openmrs/module/web/WebModuleUtil.java]: sorry but I think we need to revert this change because of my explanation above.

File: web/src/main/java/org/openmrs/web/Listener.java [hunk 6]
```diff
@@ -338,7 +339,7 @@ private String getContextPath(ServletContext servletContext) {
 	private void clearDWRFile(ServletContext servletContext) {
 		String realPath = servletContext.getRealPath("");
```
Comment (teleivo) [inline web/src/main/java/org/openmrs/web/Listener.java]: `realPath` is a string which contains the correct file separators. You can have a look at the method/interface javadocs.

the change you made would not work (same as above explanation) but if we remove `absPath` and inline this with

```java
File dwrFile = Paths.get(realPath, "WEB-INF", "dwr-modules.xml").toFile();
```

File: web/src/main/java/org/openmrs/web/Listener.java [hunk 7]
```diff
@@ -492,9 +493,7 @@ private boolean copyFile(String fromPath, String toPath) {
 	 * @param servletContext the current servlet context for the webapp
 	 */
 	public static void loadBundledModules(ServletContext servletContext) {
-		String path = servletContext.getRealPath("");
-		path += File.separator + "WEB-INF" + File.separator + "bundledModules";
-		File folder = new File(path);
+		File folder = Paths.get(servletContext.getRealPath(""),"WEB-INF","bundledModules").toFile();
```
Comment (teleivo) [inline web/src/main/java/org/openmrs/web/Listener.java]: this change is exactly what I meant when writing above comment :) great!

File: web/src/test/java/org/openmrs/module/web/WebModuleUtilTest.java [hunk 8]
```diff
@@ -218,7 +219,7 @@ public void getModuleWebFolder_shouldReturnNullIfTheDispatcherServletIsNotYetSet
 	public void getModuleWebFolder_shouldReturnTheCorrectModuleFolder() {
 		setupMocks(false);
 		String moduleId = "basicmodule";
-		String expectedPath = (REAL_PATH + "/WEB-INF/view/module/" + moduleId).replace("/", File.separator);
+		String expectedPath =Paths.get(REAL_PATH, "/WEB-INF/view/module/", moduleId).toString();
```
Comment (teleivo) [inline web/src/test/java/org/openmrs/module/web/WebModuleUtilTest.java]: you need to remove the `/` to

`Paths.get(REAL_PATH, "WEB-INF", "view", "module", moduleId).toString();`

this might fail on windows, depending on how the path is constructed on the method under test.

File: web/src/test/java/org/openmrs/module/web/WebModuleUtilTest.java [hunk 9]
```diff
@@ -231,7 +232,7 @@ public void getModuleWebFolder_shouldReturnTheCorrectModuleFolder() {
 	public void getModuleWebFolder_shouldReturnTheCorrectModuleFolderIfRealPathHasATrailingSlash() {
 		setupMocks(true);
 		String moduleId = "basicmodule";
-		String expectedPath = (REAL_PATH + "/WEB-INF/view/module/" + moduleId).replace("/", File.separator);
+		String expectedPath = Paths.get(REAL_PATH ,"/WEB-INF/view/module/", moduleId).toString();
```
Comment (teleivo) [inline web/src/test/java/org/openmrs/module/web/WebModuleUtilTest.java]: same here

**Comment by teleivo** [comment]
@chamathabeysinghe are you still working on this? You didnt respond to my comments. If not please close this PR and free up the issue on JIRA.

**Comment by teleivo** [comment]
@chamathabeysinghe I am closing this PR since there has not been any activity in a long time and we only have limited number of reviewers and need to focus our review process. Please don't feel discouraged, we really value your efforts and time you put into this! Either assign yourself to this issue again and reopen a new PR/or pick up another [issue that is ready for work](https://wiki.openmrs.org/display/docs/Getting+Started+as+a+Developer). Thanks for choosing to work on OpenMRS :)
